design/src/rfcs/rfc0012_constraint_traits.md

```diff
@@ -0,0 +1,424 @@
+[`?` operator for error propagation]: https://doc.rust-lang.org/book/ch09-02-recoverable-errors-with-result.html#a-shortcut-for-propagating-errors-the--operator
+[`std::error::Error`]: https://doc.rust-lang.org/nightly/std/error/trait.Error.html
+[`sensitive` trait]: https://awslabs.github.io/smithy/1.0/spec/core/documentation-traits.html#sensitive-trait
+[400 HTTP response message]: https://developer.mozilla.org/en-US/docs/web/http/status/400
+[`TryFrom`]: https://doc.rust-lang.org/std/convert/trait.TryFrom.html
+[`Display`]: https://doc.rust-lang.org/std/fmt/trait.Display.html
+[`Debug`]: https://doc.rust-lang.org/std/fmt/trait.Debug.html
+
+### Request deserialization
+
+We will continue to deserialize the different parts of the HTTP message into
+the regular Rust standard library types. However, just before the
+deserialization function returns, we will convert the type into the wrapper
+tuple struct that we will eventually be handed over to the operation handler.
+This is what we're _already_ doing when deserializing strings into `enums`. For
+example, given the Smithy model:
+
+```smithy
+@enum([
+    { name: "Spanish", value: "es" },
+    { name: "English", value: "en" },
+    { name: "Japanese", value: "jp" },
+])
+string Language
+```
+
+the code the client generates when deserializing a string from a JSON document
+into the `Language` enum is (excerpt):
+
+```rust
+...
+match key.to_unescaped()?.as_ref() {
+    "language" => {
+        builder = builder.set_language(
+            aws_smithy_json::deserialize::token::expect_string_or_null(
+                tokens.next(),
+            )?
+            .map(|s| {
+                s.to_unescaped()
+                    .map(|u| crate::model::Language::from(u.as_ref()))
+            })
+            .transpose()?,
+        );
+    }
+    _ => aws_smithy_json::deserialize::token::skip_value(tokens)?,
+}
+...
+```
+
+Note how the `String` gets converted to the enum via `Language::from()`.
+
+```rust
+impl std::convert::From<&str> for Language {
+    fn from(s: &str) -> Self {
+        match s {
+            "es" => Language::Spanish,
+            "en" => Language::English,
+            "jp" => Language::Japanese,
+            other => Language::Unknown(other.to_owned()),
+        }
+    }
+}
+```
+
+For constrained shapes we would do the same to parse the inner deserialized
+value into the wrapper tuple struct, except for these differences:
+
+1. For enums, the client generates an `Unknown` variant that "contains new
+   variants that have been added since this code was generated". The server
+   does not need such a variant ([#1187]).
+2. Conversions into the tuple struct are fallible (`try_from()` instead of
+   `from()`). These errors will result in a `MyStructValidationError`, wrapped
+   in the runtime type error
+   `aws_smithy_http_server::rejection::ConstraintViolation` rejection. The
```

Comment:
Will there be a way to hook these different failures for metrics so that service owners can discern between failure to parse and failure to validate in their metrics?

Reply:
Yes; the latter will be `Deserialize` rejections containing `ConstraintViolation` rejections, the former will contain other inner rejections, so it will be possible to distinguish.